### PR TITLE
[Selection syntax] Generalized autocomplete functionality

### PR DESCRIPTION
**Pull request by salazarm**
## Summary & Motivation

This PR introduces generalized autocomplete functionality that can be re-used across grammars that follow the base structure of our asset selection syntax grammar. 

As an input the autocomplete factory takes the names of attributes, their values, and the list of functions available in the grammar.  

To implement auto-complete I opted to create a new grammar so that I can allow certain things to still be valid inputs.
Benefits of separate grammar:

- `IncompleteOrExpression`, `IncompleteAndExpression`, `IncompleteFunctionCall`, etc.  allow our syntax highlighting and autocomplete visitors to visit the full expression even if parts of it are malformed. 
- DownTraversal and UpTraversal are distinguished simplifying the visitor logic
- Whitespace is differentiated between `afterExpressionWhitespace` and `afterLogicalOperatorWhitespace`, etc. greatly simplifying the visitors that provide suggestions in whitespace. (Instead of needing a visit function for every possible expression we instead have the whitespace tokens handle creating suggestions).
- `orToken`, `andToken`, `notToken` simplifies visitor logic, rather having the various expressions that contain these tokens handle what to do when the cursor is over them, we can instead have the token itself handle what happens and reduce the amount of visitor logic. 
- Separate tokens for `FunctionName` `AttributeName` `AttributeValue` - also reduces visitor logic

Generally speaking, my strategy for the grammar was to make leaf nodes that are shared among the various expression in order to allow those leaf nodes to make suggestions and reduce the number of visit functions required in the visitor (avoid combinatorial explosion of cases).

The reason for doing `this.cursorIndex - 1` in the `nodeIncludesCursor` function is that I want the suggestions to be determined by the token PRECEEDING the cursor, so I want the token immediately before the cursor to return "true" for `nodeIncludesCursor`.  

I can probably clean up the grammar a bit, but quite fatigued for this PR. In a future PR I'll look to simplify the grammar even more but what I have now passes all the tests I've written.

## How I Tested These Changes

Extensive jest testing and real testing.

Screenshot 2024-12-04 at 11 20 58 AM

### CODE REVIEW

File: python_modules/dagster-webserver/dagster_webserver/graphiql/graphiql.min.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /******/ (function() { // webpackBootstrap
-/******/ 	"use strict";
```
Comment (salazarm) [inline python_modules/dagster-webserver/dagster_webserver/graphiql/graphiql.min.js:2]: I did a grep and accidentally changed this file... I tried reverting the change with:
`git checkout origin/master -- python_modules/dagster-webserver/dagster_webserver/graphiql/graphiql.min.js` which should just use whatever is in master but I still end up with these whitespace changes... confirmed graphiql still works though

File: js_modules/dagster-ui/packages/ui-core/src/asset-selection/input/AssetSelectionLinter.ts [hunk 2]
```diff
@@ -26,5 +28,7 @@ export const lintAssetSelection = (text: string) => {
     to: CodeMirror.Pos(error.line, text.length),
   }));
 
+  console.log({lintErrors});
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/asset-selection/input/AssetSelectionLinter.ts:31]: rm

File: js_modules/dagster-ui/packages/ui-core/src/asset-selection/input/AssetSelectionLinter.ts [hunk 3]
```diff
@@ -1,23 +1,25 @@
...
+  lexer.addErrorListener(errorListener);
+
   const tokens = new CommonTokenStream(lexer);
   const parser = new AssetSelectionParser(tokens);
 
-  const errorListener = new SyntaxErrorListener();
   parser.removeErrorListeners(); // Remove default console error listener
   parser.addErrorListener(errorListener);
 
-  // Attempt to parse the input
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/asset-selection/input/AssetSelectionLinter.ts:18]: look into

File: js_modules/dagster-ui/packages/ui-core/src/selection/SelectionAutoCompleteInputParser.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      // Parsing error occurred
+      const currentErrors = errorListener.getErrors();
+
+      if (currentErrors.length > 0) {
+        const error = currentErrors[0]!;
+        const errorCharPos = error.column;
+        const errorIndex = currentPosition + errorCharPos;
+
+        // Parse up to the error
+        const validInput = input.substring(currentPosition, errorIndex);
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/selection/SelectionAutoCompleteInputParser.ts:73]: make sure this is covered by tests

**Review by hellendag — APPROVED** [review]

**Review by bengotow — APPROVED** [review]
This looks very cool 🙌  Nicely done! Cool that this can be attached to a CodeMirror parse tree visitor, I think I'd missed the implementation of that with the old AssetSelection one.

**Pull request merged**